src/navigation_plugin.cpp

```diff
@@ -68,84 +64,73 @@ void constructConstraint(std::vector<geo::Vector3>& chull, std::stringstream& co
 
-        if (r.readGroup("box", tue::config::OPTIONAL))
-        {
-            std::vector<geo::Vector3> chull; // In MAP frame
+    // Calculate cluster convex hull
+    float z_min = 1e9;
+    float z_max = -1e9;
 
-            geo::Vector3 min, max;
+    // Calculate z_min and z_max of cluster
+    std::vector<geo::Vec2f> points_2d;
+    for(std::vector<geo::Vector3>::const_iterator it = points.begin(); it != points.end(); ++it)
+    {
+        points_2d.push_back(geo::Vec2f(it->x, it->y));
 
-            if (r.readGroup("min"))
-            {
-                r.value("x", min.x);
-                r.value("y", min.y);
-                r.value("z", min.z);
-                r.endGroup();
-            }
+        z_min = std::min<float>(z_min, it->z);
+        z_max = std::max<float>(z_max, it->z);
+    }
 
-            if (r.readGroup("max"))
-            {
-                r.value("x", max.x);
-                r.value("y", max.y);
-                r.value("z", max.z);
-                r.endGroup();
-            }
+    ed::ConvexHull chull;
+    ed::convex_hull::createAbsolute(points_2d, z_min, z_max, chull);
 
-            chull.push_back(min);
-            chull.push_back(geo::Vector3(max.x, min.y, 0));
-            chull.push_back(max);
-            chull.push_back(geo::Vector3(min.x, max.y, 0));
+    constructConstraint(chull, shape_constraint);
+    return shape_constraint.str();
+}
 
-            // Transform to map frame, and add to entity constraint
-            transformChull(entity_pose, chull);
-            constructConstraint(chull, sub_shape_constraint);
+// ----------------------------------------------------------------------------------------------------
 
-            r.endGroup();
-        }
-        else if (r.readGroup("polygon", tue::config::OPTIONAL))
-        {
-            std::vector<geo::Vector3> chull; // In MAP frame
+//
+/**
+ * @brief constructCompositeShapeConstraint
+ * @param composite CompositeShape of the volume in entity frame
+ * @param entity_pose pose of the entity
+ * @return constraint string
+ */
+std::string constructCompositeShapeConstraint(geo::CompositeShapeConstPtr& composite, const geo::Pose3D& entity_pose)
+{
+    std::stringstream shape_constraint;
 
-            if (r.readArray("points", tue::config::REQUIRED))
-            {
-                while(r.nextArrayItem())
-                {
-                    double x, y;
-                    if (r.value("x", x) && r.value("y", y))
-                        chull.push_back(geo::Vector3(x, y, 0));
-                }
-                r.endArray();
-            }
-            r.endGroup();
+    bool first_sub_shape = true;
 
-            // Transform to map frame, and add to entity constraint
-            transformChull(entity_pose, chull);
-            constructConstraint(chull, sub_shape_constraint);
-        }
+    std::vector<std::pair<geo::ShapePtr, geo::Transform> > sub_shapes = composite->getShapes();
```

Comment:
Why do you make a copy? Doesn't seem needed.

Reply:
I would expect a `const std::vector<...> &sub_shapes = composite->getShapes();` to work, but you have a valid point. It looks safe, as it seems to simply return the value of the `_shape` variable. No idea what would happen if a new object would be constructed instead. I would have to consult the C++ semantics to give a verdict on that.

I can see that you may not want to modify this now, but it looks fishy that the shape code returns its own objects without `const` protection and by value rather than by reference (so the caller can decide whether to make a copy).